include/dali/core/exec/tasking/sync.h

```diff
@@ -0,0 +1,191 @@
+  /** A list of tasks waiting for this waitable object. */
+  SmallVector<WeakTask, 8> waiting_;
+
+  /** Checks whether the Waitable is ready to be acquired. */
+  virtual bool IsAcquirable() const = 0;
+
+  /** @brief Changes the state of the object to acquired (if necessary) and returns whether the
+   *         operation was successful.
+   */
+  virtual bool AcquireImpl() = 0;
+
+  /** Informs the scheduler that this Waitable is signalled and some tasks may become unblocked. */
+  void Notify(Scheduler &sched);
+
+  /** Fast comparison of weak pointers without calling `lock`. */
+  static bool task_owner_equal(const WeakTask &w, const SharedTask &t) {
+    return !w.owner_before(t) && !t.owner_before(w);
+  }
+
+  /** Checks wheter the task is waiting for this waitiable object. */
+  bool IsWaitedForBy(const SharedTask &task) const {
+    auto it = std::find_if(waiting_.begin(), waiting_.end(),
+                           [&](auto &t) { return task_owner_equal(t, task); });
+    return it != waiting_.end();
+  }
+
+  /** Tries to acquire the waitable object on behalf of a task.
+   *
+   * This function can fail for two reasons:
+   * - the task is not waiting for this waitable
+   * - AcquireImpl fails
+   */
+  bool TryAcquire(const SharedTask &task) {
+    auto it = std::find_if(waiting_.begin(), waiting_.end(),
+                           [&](auto &t) { return task_owner_equal(t, task); });
+    if (it == waiting_.end())
+      return false;
+    if (AcquireImpl()) {
+      waiting_.erase(it);
+      return true;
+    }
+    return false;
+  }
+
+  /** Adds a task to the waiting list if it's not already there.
+   *
+   * @return `true`, if the task was added; `false` if it was already on the list.
+   */
+  bool AddToWaiting(const SharedTask &task) {
+    if (IsWaitedForBy(task))
+      return false;
+    waiting_.push_back(task);
+    return true;
+  }
+
+ public:
+  virtual ~Waitable() = default;
+};
+
+/** A waitable object that changes the state at most once, from incomplete to complete.
+ *
+ * A completion event is any Waitable which cannot become "unsignalled" - for example, once
+ * a task is complete it will ever remain so, it cannot go back to the incomplete state.
+ */
+class CompletionEvent : public Waitable {
+ protected:
+  bool AcquireImpl() override {
+    // Nothing to acquire - just return true if the event is completed.
+    return IsAcquirable();
+  }
+
+  void MarkAsComplete() {
+    completed_ = true;
+  }
+
+  bool IsAcquirable() const override {
+    return completed_;
+  }
+
+ private:
+  bool completed_ = false;
+};
+
+/** A waitable object which has a Release method, which can be called from outside any task.
+ */
+class Releasable : public Waitable {
+ public:
+  void Release(Scheduler &sched) {
+    ReleaseImpl();
+    Notify(sched);
+  }
+
+ protected:
+  /** Changes the internal state of the object.
+   *
+   * If IsAcquirable is called atomically after a successful ReleaseImpl, then it must return true.
```

Comment:
```suggestion
   * If IsAcquirable is called atomically after a successful ReleaseImpl.
```

Reply:
If **IsAcquirable** is called atomically after a successful ReleaseImpl, then **it** must return true.
It refers to IsAcquirable and it, indeed, must return true after ReleaseImpl and IsAcquirable are invoked atomically.

Reply:
Rewritten.